### PR TITLE
CurverBs:Add size of chunkfilepool to optimize whether the logical po…

### PR DESCRIPTION
**Pull request by yubindy**
Signed-off-by: Zhao ZeYu <2302370185@qq.com>

### What problem does this PR solve?

Issue Number: #1058 

Problem Summary:

### What is changed and how it works?

What's Changed:add chunkfilepool size at topology_stat to optimize the logical pool is allocated

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by yuxubinchen** [comment]
recheck

File: curvefs/test/mds/mock/mock_topology.h [hunk 1]
```diff
@@ -288,7 +288,7 @@ class MockTopology : public TopologyImpl {
     MOCK_CONST_METHOD2(GetServerByHostName,
                        bool(const std::string &hostName, Server *out));
     MOCK_CONST_METHOD3(GetServerByHostIpPort, bool(const std::string &hostIp,
-                                                   uint32_t port, Server *out));
+                uint32_t port, Server *out));
```
Comment (xu-chaojie) [inline curvefs/test/mds/mock/mock_topology.h:291]: Do not introduce changes unrelated to the issue

File: src/mds/topology/topology.cpp [hunk 2]
```diff
@@ -679,6 +680,14 @@ bool TopologyImpl::GetServer(ServerIdType serverId, Server *out) const {
         *out = it->second;
         return true;
     }
+    for (auto po : serverMap_) {
```
Comment (cw123) [inline src/mds/topology/topology.cpp:683]: change po to a meaningful name

Comment (xu-chaojie) [inline src/mds/topology/topology.cpp:683]: Do not introduce changes unrelated to the issue

File: src/mds/topology/topology.cpp [hunk 3]
```diff
@@ -111,6 +111,7 @@ int TopologyImpl::AddZone(const Zone &data) {
 int TopologyImpl::AddServer(const Server &data) {
     ReadLockGuard rlockZone(zoneMutex_);
     WriteLockGuard wlockServer(serverMutex_);
+    LOG(INFO) << "AddServer: " << data.GetId();
```
Comment (xu-chaojie) [inline src/mds/topology/topology.cpp:114]: make this log more clear

**Comment by cw123** [comment]
Combine 3 commits into one commit

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 4]
```diff
@@ -175,9 +175,12 @@ bool TopologyChunkAllocatorImpl::ChooseSingleLogicalPool(
         if (!topology_->GetPhysicalPool(lPool.GetPhysicalPoolId(), &pPool)) {
             continue;
         }
-        uint64_t diskCapacity = pPool.GetDiskCapacity();
-        // calculate actual capacity available
-        diskCapacity = diskCapacity * poolUsagePercentLimit_ / 100;
```
Comment (xu-chaojie) [inline src/mds/topology/topology_chunk_allocator.cpp:180]: poolUsagePercentLimit_ is required, indicating the percentage of chunkfilepool usage

File: src/chunkserver/register.h [hunk 5]
```diff
@@ -48,6 +48,7 @@ struct RegisterOptions {
     std::string chunkserverDiskType;
     int registerRetries;
     int registerTimeout;
+    uint64_t chunkfilepoolsize;
```
Comment (cw123) [inline src/chunkserver/register.h:51]: use Camel-Case

File: proto/heartbeat.proto [hunk 6]
```diff
@@ -87,6 +87,8 @@ message ChunkServerStatisticInfo {
     required uint64 chunkSizeLeftBytes = 6;
     // 回收站中chunk占用的磁盘空间
     required uint64 chunkSizeTrashedBytes = 7;
+   // chunkfilepool的大小
+    required uint64 chunkfilepoolSize = 8;
```
Comment (cw123) [inline proto/heartbeat.proto:91]: use Camel-Case

File: src/mds/topology/topology_stat.h [hunk 7]
```diff
@@ -83,6 +83,8 @@ struct ChunkServerStat {
     uint64_t chunkSizeLeftBytes;
     // Size of chunks in recycle bin
     uint64_t chunkSizeTrashedBytes;
+    // Size of chunkfilepool
+    uint64_t chunkfilepoolSize;
```
Comment (cw123) [inline src/mds/topology/topology_stat.h:87]: use Camel-Case

File: src/mds/topology/topology_stat.h [hunk 8]
```diff
@@ -123,6 +125,14 @@ class TopologyStat {
...
     virtual bool GetChunkServerStat(ChunkServerIdType csId,
         ChunkServerStat *stat) = 0;
+    /**
+     * @brief fetch the statistic information of chunkPool size that sent by heartbeat
+     *
+     * @param pId physicalId
+     * @param chunkpoolsize the size of chunkpool
+     */    
+    virtual bool GetChunkPoolSize(PoolIdType pId,
+    uint64_t *chunkpoolsize) = 0;
```
Comment (cw123) [inline src/mds/topology/topology_stat.h:135]: Camel-Case

File: src/mds/topology/topology_stat.cpp [hunk 9]
```diff
@@ -58,6 +65,17 @@ bool TopologyStatImpl::GetChunkServerStat(ChunkServerIdType csId,
...
+bool TopologyStatImpl::GetChunkPoolSize(PoolIdType pId,
+    uint64_t *chunkpoolsize) {
+    ReadLockGuard rLock(statsLock_);
+    auto it = ChunkPoolSize_.find(pId);
+    if (it != ChunkPoolSize_.end()) {
+        *chunkpoolsize = it->second;
+        return true;
+    }
+    return false;
+    }
```
Comment (cw123) [inline src/mds/topology/topology_stat.cpp:77]: delete spaces

File: src/mds/topology/topology_stat.cpp [hunk 10]
```diff
@@ -38,11 +39,17 @@ namespace topology {
 void TopologyStatImpl::UpdateChunkServerStat(ChunkServerIdType csId,
     const ChunkServerStat &stat) {
     WriteLockGuard wLock(statsLock_);
+
+    PoolIdType belongPhysicalPoolId = UNINTIALIZE_ID;
+    int ret = topo_-> GetBelongPhysicalPoolId(csId, &belongPhysicalPoolId);
```
Comment (cw123) [inline src/mds/topology/topology_stat.cpp:44]: Judgment return value

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
这个日志很奇怪，看起来测试点是过了但是，会一直循环跑下去，直到超时1h

**Comment by yuxubinchen** [comment]
@cw123

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
根据chunkserve conf的配置，如果使用chunkfielpool，mds使用chunkfilepool的size,否则使用原来的物理容量.

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 11]
```diff
@@ -188,9 +188,13 @@ bool TopologyChunkAllocatorImpl::ChooseSingleLogicalPool(
...
         // calculate remaining capacity
-        uint64_t diskRemainning =
-            (diskCapacity > alloc) ? diskCapacity - alloc : 0;
-
+        if (diskCapacity < alloc) {
+            continue;
+        }
+        uint64_t diskRemainning;
+        if (!topoStat_->GetChunkPoolSize(pPool.GetId(), &diskRemainning)) {
+            continue;
```
Comment (xu-chaojie) [inline src/mds/topology/topology_chunk_allocator.cpp:196]: 这个地方的逻辑是不对的，总的空间是chunkfilepool的空间，使用掉的空间是alloc的空间，剩余空间应该是两者相减，再通过poolUsagePercentLimit_这个字段来预留一部分可分配空间防止failover时，空间溢出

Comment (xu-chaojie) [reply src/mds/topology/topology_chunk_allocator.cpp:196]: 整个思路再给你理一下，分为几种情况：
1. 不使用chunkfilepool的时候：
此时，总量应该是磁盘的全部空间，分配量是alloc的空间，可用空间是两者相减，但是需要预留两种空间，一是raft WAL的空间，而是failover的空间
2. 使用chunkfilepool，不使用walpool的情况：
这时，总量是chunkfilepool的总空间，分配量是alloc的空间，可用空间是两者相减，但是需要预留failover的空间
3. 使用chunkfilepool，并且使用walpool的情况：
与第2种情况一样：
4.  使用chunkfilepool，walpool共用chunkfilepool的情况：
这时，总量是chunkfilepool的总空间，分配量是alloc的空间，可用空间时两者相减，但是需要预留两个空间，一是raft WAL的空间，而是failover的空间

Comment (yuxubinchen) [reply src/mds/topology/topology_chunk_allocator.cpp:196]: Ok thanks, I will fix it, I originally thought that the size of the chunkfilepool is the size of the available space. So no subtracting alloc. I get it now

Comment (yubindy) [reply src/mds/topology/topology_chunk_allocator.cpp:196]: Do you need any strategy for reservation? The feasible method for me now is to use the size of chunkfilePool minus the size of walPool and the size of copyset that have been used now. I think it should be extra reservation, which may be reserved according to the percentage of the remaining size. Stay, no, I don't have any better ideas.
对于预留需要什么策略吗，我现在可行的方法是使用chunkfilePool的size减去现在已经使用walPool的size和copyset的size,我觉得是应该是需要进行额外预留，可能可以按照剩下size的百分比预留，否侧我没有什么比较好的想法了

Comment (yubindy) [reply src/mds/topology/topology_chunk_allocator.cpp:196]: 而对于walPool的size和copyset的size,因为共用一个chunkfilepool 的接口，我能想到统计size的是用walpool的目录下的文件个数*segment_size

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
image

**Comment by yubindy** [comment]
image
在快照这里一直显示是创建的文件比剩余的chunkfilepool的size大,所以一直失败，而我在增加日志确定后确实是比chunkfilepool的size大，所以会一直失败

**Comment by yubindy** [comment]
image

**Comment by yubindy** [comment]
image

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

File: conf/mds.conf [hunk 12]
```diff
@@ -163,6 +163,10 @@ mds.topology.PoolUsagePercentLimit=85
 mds.topology.choosePoolPolicy=0
 # enable LogicalPool ALLOW/DENY status
 mds.topology.enableLogicalPoolStatus=false
+# 当使用 chunkfilepool 进行分配时需预留failover空间，
+#建议每个chunkserver预留8%的预留空间，
+#即为 mds.chunkserver.failure.tolerance*8
+mds.topology.ChunkFilePoolfailoverReserve=24
```
Comment (xu-chaojie) [inline conf/mds.conf:169]: 一个zone 标准配置20台节点，允许3台节点failover，那么剩余17台机器需要承载原先20台机器的空间，那么20台机器每台最多占多少空间？

File: src/chunkserver/chunkserver.cpp [hunk 13]
```diff
@@ -150,6 +150,7 @@ int ChunkServer::Run(int argc, char** argv) {
     std::string raftLogProtocol = UriParser::GetProtocolFromUri(raftLogUri);
     std::shared_ptr<FilePool> walFilePool = nullptr;
     bool useChunkFilePoolAsWalPool = true;
+    uint32_t useChunkFilePoolAsWalPoolReserve = 15;
```
Comment (xu-chaojie) [inline src/chunkserver/chunkserver.cpp:153]: Shouldn't this come from the config file?

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
image

File: conf/mds.conf [hunk 14]
```diff
@@ -163,6 +163,11 @@ mds.topology.PoolUsagePercentLimit=85
 mds.topology.choosePoolPolicy=0
 # enable LogicalPool ALLOW/DENY status
 mds.topology.enableLogicalPoolStatus=false
+# 当使用 chunkfilepool 进行分配时需预留failover空间，
+#和mds.chunkserver.failure.tolerance设置有关
+#一个zone 标准配置20台节点，如果允许3台节点failover，
+#那么剩余17台机器需要承载原先20台机器的空间,17/20=0.85,
+mds.topology.ChunkFilePoolfailoverReserve=15
```
Comment (xu-chaojie) [inline conf/mds.conf:170]: remove “mds.topology.PoolUsagePercentLimit ” when it not needed

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 15]
```diff
@@ -198,9 +198,17 @@ void TopologyChunkAllocatorImpl::GetRemainingSpaceInLogicalPool(
...
+        uint64_t diskCapacity = 0;
+        if (chunkFilePoolAllocHelp_->getUseChunkFilepool()) {
+            topoStat_->GetChunkPoolSize(lPool.GetPhysicalPoolId(),
+                        &diskCapacity);
+            available = chunkFilePoolAllocHelp_->getAvailable();
+            diskCapacity = diskCapacity * available / 100;
+        } else {
+            diskCapacity = pPool.GetDiskCapacity();
+            // calculate actual capacity available
+            diskCapacity = diskCapacity * available / 100;
```
Comment (xu-chaojie) [inline src/mds/topology/topology_chunk_allocator.cpp:210]: available here has no assignment？

File: src/mds/topology/topology_chunk_allocator.h [hunk 16]
```diff
@@ -45,6 +46,45 @@ enum class ChoosePoolPolicy {
...
+            std::memory_order_release);
+        ChunkFilePoolPoolWalReserve.store(useWalasChunkfilePoolReserve_,
+            std::memory_order_release);
+    }
+    bool getUseChunkFilepool() {
+        return useChunkFilepool.load(std::memory_order_acquire);
+    }
+    // After removing the reserved space, the remaining percentage
+    uint32_t getAvailable() {
+        if (useWalasChunkFilePool.load(std::memory_order_acquire)) {
```
Comment (xu-chaojie) [inline src/mds/topology/topology_chunk_allocator.h:68]: use Wal as ChunkFilePool   ？

File: proto/topology.proto [hunk 17]
```diff
@@ -154,6 +154,9 @@ message ChunkServerRegistRequest {
     // already registed chunkserver should take chunkserverId & token
     optional uint32 chunkServerID = 6;
     optional string token = 7;
+    required uint64 chunkFilepoolSize = 8;
```
Comment (xu-chaojie) [inline proto/topology.proto:157]: use optional for backward compatibility

File: proto/heartbeat.proto [hunk 18]
```diff
@@ -87,6 +87,8 @@ message ChunkServerStatisticInfo {
     required uint64 chunkSizeLeftBytes = 6;
     // 回收站中chunk占用的磁盘空间
     required uint64 chunkSizeTrashedBytes = 7;
+   // chunkfilepool的大小
+    required uint64 chunkFilepoolSize = 8;
```
Comment (xu-chaojie) [inline proto/heartbeat.proto:91]: use optional for backward compatibility

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 19]
```diff
@@ -198,9 +198,17 @@ void TopologyChunkAllocatorImpl::GetRemainingSpaceInLogicalPool(
...
             continue;
         }
-        uint64_t diskCapacity = pPool.GetDiskCapacity();
-        // calculate actual capacity available
-        diskCapacity = diskCapacity * poolUsagePercentLimit_ / 100;
+        uint64_t diskCapacity = 0;
+        if (chunkFilePoolAllocHelp_->getUseChunkFilepool()) {
+            topoStat_->GetChunkPoolSize(lPool.GetPhysicalPoolId(),
+                        &diskCapacity);
+            available -= chunkFilePoolAllocHelp_->getAvailable();
```
Comment (xu-chaojie) [inline src/mds/topology/topology_chunk_allocator.cpp:205]: why use -= ？

**Review by xu-chaojie — CHANGES_REQUESTED** [review]

File: src/mds/topology/topology_chunk_allocator.h [hunk 20]
```diff
@@ -45,6 +46,43 @@ enum class ChoosePoolPolicy {
     kWeight,
 };
 
+class ChunkFilePoolAllochelp {
```
Comment (cw123) [inline src/mds/topology/topology_chunk_allocator.h:49]: ChunkFilePoolAllochelp -> ChunkFilePoolAllocHelp

File: src/mds/topology/topology_chunk_allocator.h [hunk 21]
```diff
@@ -136,8 +188,16 @@ class TopologyChunkAllocatorImpl : public TopologyChunkAllocator {
     std::shared_ptr<AllocStatistic> allocStatistic_;
 
     // usage limit of pool
-    uint32_t poolUsagePercentLimit_;
+    uint32_t available;
```
Comment (cw123) [inline src/mds/topology/topology_chunk_allocator.h:191]: What is the unit, is it bytes or a percentage？
add the comment

File: src/mds/topology/topology_service_manager.cpp [hunk 22]
```diff
@@ -61,6 +61,21 @@ void TopologyServiceManager::RegistChunkServer(
...
+    ChunkServerStat stat;
+    bool useChunkFilePoolAsWalPool = false;
+    uint32_t useChunkFilePoolAsWalPoolReserve = 15;
+    bool UseChunkFilepool = false;
+    if (request->has_chunkfilepoolsize()) {
+        stat.chunkFilepoolSize = request->chunkfilepoolsize();
+    }
+    if (request->has_usechunkfilepoolaswalpool()) {
+        UseChunkFilepool = true;
+        useChunkFilePoolAsWalPool = request->usechunkfilepoolaswalpool();
```
Comment (cw123) [inline src/mds/topology/topology_service_manager.cpp:73]: Check if this field is set before use

File: src/mds/topology/topology_service_manager.cpp [hunk 23]
```diff
@@ -61,6 +61,21 @@ void TopologyServiceManager::RegistChunkServer(
...
+        useChunkFilePoolAsWalPoolReserve =
```
Comment (cw123) [inline src/mds/topology/topology_service_manager.cpp:74]: Check if this field is set before use

**Comment by yubindy** [comment]
recheck

File: src/chunkserver/register.cpp [hunk 24]
```diff
@@ -64,6 +64,14 @@ int Register::RegisterToMDS(const ChunkServerMetadata *localMetadata,
         req.set_externalip(ops_.chunkserverExternalIp);
     }
     req.set_port(ops_.chunkserverPort);
+    uint64_t chunkPoolSize = ops_.chunkFilepool->Size() *
+                ops_.chunkFilepool->GetFilePoolOpt().fileSize;
+    req.set_chunkfilepoolsize(chunkPoolSize);
+    if (ops_.chunkFilepool->GetFilePoolOpt().getFileFromPool) {
+    req.set_usechunkfilepoolaswalpool(ops_.useChunkFilePoolAsWalPool);
+    req.set_usechunkfilepoolaswalpoolreserve(
```
Comment (xu-chaojie) [inline src/chunkserver/register.cpp:72]: indentation

File: src/mds/topology/topology_service_manager.cpp [hunk 25]
```diff
@@ -61,6 +61,21 @@ void TopologyServiceManager::RegistChunkServer(
     ChunkServerRegistResponse *response) {
     std::string hostIp = request->hostip();
     uint32_t port = request->port();
+    ChunkServerStat stat;
+    bool useChunkFilePoolAsWalPool = false;
+    uint32_t useChunkFilePoolAsWalPoolReserve = 15;
+    bool UseChunkFilepool = false;
+    if (request->has_chunkfilepoolsize()) {
```
Comment (cw123) [inline src/mds/topology/topology_service_manager.cpp:68]: 考虑一下，这里的几个变量。如果注册的时候，request没有设置，是否要继续注册。
If the request is not set during registration, whether to continue the registration.

File: src/mds/topology/topology_chunk_allocator.h [hunk 26]
```diff
@@ -135,9 +187,17 @@ class TopologyChunkAllocatorImpl : public TopologyChunkAllocator {
...
 
+    /**
+     * @brief topology statistic module
+     */
+    std::shared_ptr<TopologyStat> topoStat_;
+    /**
+     * @brief chunkfilepool help with capacity allocation
+     */
+    std::shared_ptr<ChunkFilePoolAllocHelp> ChunkFilePoolAllocHelp_;
```
Comment (cw123) [inline src/mds/topology/topology_chunk_allocator.h:200]: 类定义使用以大写字母开头；成员变量以小写字母开头，加上_后缀；函数以大写字母开头；函数的入参以小写字母开头，后面不加_后缀。新加的代码按照这种风格统一一下，就不一一指出了。
Class definitions start with an uppercase letter; member variables start with a lowercase letter and _ suffix; functions start with an uppercase letter; function input parameters start with a lowercase letter, without the _ suffix.

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 27]
```diff
@@ -198,9 +198,17 @@ void TopologyChunkAllocatorImpl::GetRemainingSpaceInLogicalPool(
...
             continue;
         }
-        uint64_t diskCapacity = pPool.GetDiskCapacity();
-        // calculate actual capacity available
-        diskCapacity = diskCapacity * poolUsagePercentLimit_ / 100;
+        uint64_t diskCapacity = 0;
+        if (ChunkFilePoolAllocHelp_->getUseChunkFilepool()) {
+            topoStat_->GetChunkPoolSize(lPool.GetPhysicalPoolId(),
+                        &diskCapacity);
+            available *= ChunkFilePoolAllocHelp_->getAvailable();
```
Comment (cw123) [inline src/mds/topology/topology_chunk_allocator.cpp:205]: 如果GetRemainingSpaceInLogicalPool执行多次，这里的available是不是会越变越小，每走到一次这个分支，都会改变available的值。
这个逻辑增加一个单元测试，在使用chunk file pool的场景下，多次执行GetRemainingSpaceInLogicalPool，判断获取的remaining space是否正确。

If GetRemainingSpaceInLogicalPool is executed multiple times, will the available value here become smaller and smaller? Every time this branch is reached, the value of available will be changed.
This logic adds a unit test. In the scenario of using chunk file pool, execute GetRemainingSpaceInLogicalPool multiple times to determine whether the acquired remaining space is correct.

**Review by xu-chaojie — APPROVED** [review]

**Review by cw123 — APPROVED** [review]

**Pull request merged**